File: nautilus_trader/adapters/longport/factories.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------

"""
Longport adapter factories.

Provides factory classes for creating Longport data and execution clients.
"""

import asyncio
from functools import lru_cache

from nautilus_trader.adapters.longport.config import (
    LongportDataClientConfig,
    LongportExecClientConfig,
)
from nautilus_trader.adapters.longport.data import LongportDataClient
from nautilus_trader.adapters.longport.execution import LongportExecutionClient
from nautilus_trader.adapters.longport.providers import LongportInstrumentProvider
from nautilus_trader.cache.cache import Cache
from nautilus_trader.common.component import LiveClock
from nautilus_trader.common.component import MessageBus
from nautilus_trader.config import InstrumentProviderConfig
from nautilus_trader.core import nautilus_pyo3
from nautilus_trader.live.factories import LiveDataClientFactory
from nautilus_trader.live.factories import LiveExecClientFactory


@lru_cache(1)
def get_cached_longport_clients(
    app_key: str | None = None,
    app_secret: str | None = None,
    access_token: str | None = None,
    http_url: str | None = None,
    quote_ws_url: str | None = None,
    trade_ws_url: str | None = None,
) -> tuple:
    """
    Cache and return Longport QuoteContext, HTTP client, WebSocket client, and event receiver.

    Following the OKX architecture pattern, QuoteContext is created once and shared
    between HTTP and WebSocket clients via Arc.

    If a cached client with matching parameters already exists, the cached client will be returned.

    Parameters
    ----------
    app_key : str, optional
        The app key for the client.
    app_secret : str, optional
        The app secret for the client.
    access_token : str, optional
        The access token for the client.
    http_url : str, optional
        Custom HTTP URL (default: https://openapi.longportapp.com)
    quote_ws_url : str, optional
        Custom quote WebSocket URL (default: wss://openapi-quote.longportapp.com/v2)
    trade_ws_url : str, optional
        Custom trade WebSocket URL (default: wss://openapi-trade.longportapp.com/v2)

    Returns
    -------
    tuple
        (quote_ctx, http_client, ws_client, event_receiver)

    """
    quote_ctx, event_receiver = nautilus_pyo3.longport.create_quote_context(
        app_key=app_key or "",
        app_secret=app_secret or "",
        access_token=access_token or "",
        http_url=http_url,
        quote_ws_url=quote_ws_url,
        trade_ws_url=trade_ws_url,
    )

    http_client = nautilus_pyo3.longport.LongportHttpClient(quote_ctx)
    ws_client = nautilus_pyo3.longport.LongportWebSocketClient(quote_ctx)

    return quote_ctx, http_client, ws_client, event_receiver


@lru_cache(1)
def get_cached_longport_instrument_provider(
    markets: tuple[str, ...],
    config: InstrumentProviderConfig | None = None,
) -> LongportInstrumentProvider:
    """
    Cache and return a Longport instrument provider.

    If a cached provider already exists, then that provider will be returned.

    Parameters
    ----------
    markets : tuple[str, ...]
        The markets to load instruments for.
    config : InstrumentProviderConfig, optional
        The instrument provider configuration.

    Returns
    -------
    LongportInstrumentProvider

    """
    return LongportInstrumentProvider(
        markets=markets,
        config=config,
    )


class LongportLiveDataClientFactory(LiveDataClientFactory):
    """
    Provides a Longport live data client factory using Rust implementation.
    """

    @staticmethod
    def create(  # type: ignore
        loop: asyncio.AbstractEventLoop,
        name: str,
        config: LongportDataClientConfig,
        msgbus: MessageBus,
        cache: Cache,
        clock: LiveClock,
    ):
        """
        Create a new Longport data client using the Rust implementation.

        Parameters
        ----------
        loop : asyncio.AbstractEventLoop
            The event loop for the client.
        name : str
            The custom client ID.
        config : LongportDataClientConfig
            The client configuration.
        msgbus : MessageBus
            The message bus for the client.
        cache : Cache
            The cache for the client.
        clock : LiveClock
            The clock for the client.

        Returns
        -------
        LongportDataClient
            The Python Longport data client (which wraps Rust implementation).

        """
        # Get the instrument provider
        provider = get_cached_longport_instrument_provider(
            markets=tuple(config.markets),
            config=config.instrument_provider,
        )

        # Create and return the Python LongportDataClient
        # The client will create QuoteContext, HTTP/WebSocket clients internally in _connect()
        client = LongportDataClient(
            loop=loop,
            msgbus=msgbus,
            cache=cache,
            clock=clock,
            instrument_provider=provider,
            config=config,
            name=name,
        )

        return client


class LongportLiveExecClientFactory(LiveExecClientFactory):
    """
    Provides a Longport live execution client factory.
    """

    @staticmethod
    def create(  # type: ignore
        loop: asyncio.AbstractEventLoop,
        name: str,
        config: LongportExecClientConfig,
        msgbus: MessageBus,
        cache: Cache,
        clock: LiveClock,
    ) -> LongportExecutionClient:
        """
        Create a new Longport execution client.

        Parameters
        ----------
        loop : asyncio.AbstractEventLoop
            The event loop for the client.
        name : str
            The custom client ID.
        config : LongportExecClientConfig
            The client configuration.
        msgbus : MessageBus
            The message bus for the client.
        cache : Cache
            The cache for the client.
        clock : LiveClock
            The clock for the client.

        Returns
        -------
        LongportExecutionClient

        """
        provider = get_cached_longport_instrument_provider(
            markets=tuple(config.markets),
            config=config.instrument_provider,
        )
        return LongportExecutionClient(
            loop=loop,
            msgbus=msgbus,
            cache=cache,
            clock=clock,
            instrument_provider=provider,
            config=config,
            name=name,
        )

File: crates/adapters/longport/src/factories.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Factory functions for creating Longport clients and components.

use std::{any::Any, cell::RefCell, rc::Rc};

use nautilus_common::{cache::Cache, clients::ExecutionClient};
use nautilus_execution::client::core::ExecutionClientCore;
use nautilus_model::{
    enums::{AccountType, OmsType},
    identifiers::ClientId,
};
use nautilus_system::factories::{ClientConfig, ExecutionClientFactory};

use crate::{
    common::consts::LONGPORT_VENUE,
    config::{LongportDataClientConfig, LongportExecClientConfig},
    execution::LongportExecutionClient,
};

impl ClientConfig for LongportDataClientConfig {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

impl ClientConfig for LongportExecClientConfig {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

// NOTE: LongportDataClientFactory is removed since DataClient is now implemented in Python
// following the OKX architecture pattern. Only ExecutionClientFactory remains in Rust.

/// Factory for creating Longport execution clients.
#[derive(Debug, Clone)]
#[cfg_attr(
    feature = "python",
    pyo3::pyclass(module = "nautilus_trader.core.nautilus_pyo3.longport")
)]
#[cfg_attr(
    feature = "python",
    pyo3_stub_gen::derive::gen_stub_pyclass(module = "nautilus_trader.adapters.longport")
)]
pub struct LongportExecutionClientFactory;

impl LongportExecutionClientFactory {
    /// Creates a new [`LongportExecutionClientFactory`] instance.
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Default for LongportExecutionClientFactory {
    fn default() -> Self {
        Self::new()
    }
}

impl ExecutionClientFactory for LongportExecutionClientFactory {
    fn create(
        &self,
        name: &str,
        config: &dyn ClientConfig,
        cache: Rc<RefCell<Cache>>,
    ) -> anyhow::Result<Box<dyn ExecutionClient>> {
        let longport_config = config
            .as_any()
            .downcast_ref::<LongportExecClientConfig>()
            .ok_or_else(|| {
                anyhow::anyhow!(
                    "Invalid config type for LongportExecutionClientFactory. Expected LongportExecClientConfig, was {config:?}",
                )
            })?
            .clone();

        // Longport uses cash accounts by default (spot trading)
        let account_type = AccountType::Cash;
        let oms_type = OmsType::Hedging;

        let core = ExecutionClientCore::new(
            longport_config.trader_id,
            ClientId::from(name),
            *LONGPORT_VENUE,
            oms_type,
            longport_config.account_id,
            account_type,
            None, // base_currency
            cache,
        );

        let client = LongportExecutionClient::new(core, longport_config)?;

        Ok(Box::new(client))
    }

    fn name(&self) -> &'static str {
        "LONGPORT"
    }

    fn config_type(&self) -> &'static str {
        "LongportExecClientConfig"
    }
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, rc::Rc};

    use nautilus_common::cache::Cache;
    use nautilus_model::identifiers::{AccountId, ClientId, TraderId};
    use rstest::rstest;

    use super::*;
    use crate::common::enums::LongportMarket;

    #[rstest]
    fn test_longport_execution_client_factory_creation() {
        let factory = LongportExecutionClientFactory::new();
        assert_eq!(factory.name(), "LONGPORT");
        assert_eq!(factory.config_type(), "LongportExecClientConfig");
    }

    #[rstest]
    fn test_longport_execution_client_factory_default() {
        let factory = LongportExecutionClientFactory::new();
        assert_eq!(factory.name(), "LONGPORT");
    }

    #[rstest]
    fn test_longport_exec_client_config_implements_client_config() {
        let config = LongportExecClientConfig {
            trader_id: TraderId::from("TRADER-001"),
            account_id: AccountId::from("LONGPORT-001"),
            app_key: Some("test_key".to_string()),
            app_secret: Some("test_secret".to_string()),
            access_token: Some("test_token".to_string()),
            markets: vec![LongportMarket::HK],
            ..Default::default()
        };

        let boxed_config: Box<dyn ClientConfig> = Box::new(config);
        let downcasted = boxed_config.as_any().downcast_ref::<LongportExecClientConfig>();

        assert!(downcasted.is_some());
    }

    #[rstest]
    fn test_longport_execution_client_factory_creates_client() {
        let factory = LongportExecutionClientFactory::new();
        let config = LongportExecClientConfig {
            trader_id: TraderId::from("TRADER-001"),
            account_id: AccountId::from("LONGPORT-001"),
            app_key: Some("test_key".to_string()),
            app_secret: Some("test_secret".to_string()),
            access_token: Some("test_token".to_string()),
            markets: vec![LongportMarket::HK, LongportMarket::US],
            ..Default::default()
        };

        let cache = Rc::new(RefCell::new(Cache::default()));

        let result = factory.create("LONGPORT-TEST", &config, cache);

        // The test will fail because Longport SDK validates the token
        // This is expected behavior - the factory correctly attempts to create the client
        // and Longport SDK rejects invalid credentials
        match result {
            Ok(client) => {
                // If somehow it succeeds (e.g., mocking), verify basic properties
                assert_eq!(client.client_id(), ClientId::from("LONGPORT-TEST"));
                assert_eq!(client.venue(), *LONGPORT_VENUE);
            }
            Err(e) => {
                // Expected: Longport SDK rejects invalid test credentials
                // Verify the error is related to authentication or connection
                let error_msg = e.to_string().to_lowercase();
                assert!(
                    error_msg.contains("token") ||
                    error_msg.contains("auth") ||
                    error_msg.contains("credential") ||
                    error_msg.contains("trade context") ||
                    error_msg.contains("invalid"),
                    "Expected authentication/connection error, got: {e}"
                );
            }
        }
    }

    #[rstest]
    fn test_longport_execution_client_factory_rejects_wrong_config_type() {
        let factory = LongportExecutionClientFactory::new();
        let wrong_config = LongportDataClientConfig::default();

        let cache = Rc::new(RefCell::new(Cache::default()));

        let result = factory.create("LONGPORT-TEST", &wrong_config, cache);
        assert!(result.is_err());
        assert!(
            result
                .err()
                .unwrap()
                .to_string()
                .contains("Invalid config type")
        );
    }

    #[rstest]
    fn test_longport_data_client_config_implements_client_config() {
        let config = LongportDataClientConfig {
            app_key: Some("test_key".to_string()),
            app_secret: Some("test_secret".to_string()),
            access_token: Some("test_token".to_string()),
            markets: vec![LongportMarket::US],
            ..Default::default()
        };

        let boxed_config: Box<dyn ClientConfig> = Box::new(config);
        let downcasted = boxed_config.as_any().downcast_ref::<LongportDataClientConfig>();

        assert!(downcasted.is_some());
    }
}
